crates/symbolicator-service/src/download/index.rs

```diff
@@ -0,0 +1,244 @@
+use std::fmt::Write;
+use std::fs::File;
+use std::io::BufReader;
+use std::sync::Arc;
+use std::time::Duration;
+
+use futures::future::BoxFuture;
+use futures::prelude::*;
+use futures::stream::FuturesUnordered;
+use symbolic::common::{AccessPattern, ByteView};
+use symbolicator_sources::{HttpRemoteFile, HttpSourceConfig, SourceLocation, SymstoreIndex};
+use tempfile::NamedTempFile;
+use url::Url;
+
+use crate::caches::versions::SYMSTORE_INDEX_VERSIONS;
+use crate::caches::CacheVersions;
+use crate::caching::{Cache, CacheContents, CacheItemRequest, CacheKey, Cacher, SharedCacheRef};
+use crate::types::Scope;
+
+use super::DownloadService;
+
+#[derive(Debug, Clone)]
+struct FetchSymstoreIndexSegment {
+    segment_id: u32,
+    source: Arc<HttpSourceConfig>,
+    downloader: Arc<DownloadService>,
+}
+
+#[derive(Debug, Clone)]
+struct FetchSymstoreIndex {
+    scope: Scope,
+    last_id: u32,
+    segment_cache: Arc<Cacher<FetchSymstoreIndexSegment>>,
+    source: Arc<HttpSourceConfig>,
+    downloader: Arc<DownloadService>,
+}
+
+#[tracing::instrument(skip(downloader, source), fields(source.url = %source.url), err)]
+async fn download_index_segment(
+    downloader: Arc<DownloadService>,
+    source: Arc<HttpSourceConfig>,
+    segment: u32,
+    file: &mut File,
+) -> CacheContents {
+    let loc = format!("000Admin/{segment:0>10}");
+    let remote_file = HttpRemoteFile::new(Arc::clone(&source), SourceLocation::new(loc));
+    let temp_file = NamedTempFile::new()?;
+
+    tracing::debug!("Downloading index segment");
+
+    downloader
+        .download(remote_file.into(), temp_file.path().to_path_buf())
+        .await?;
+
+    let buf = BufReader::new(temp_file);
+    let mut index = SymstoreIndex::default();
+    index.extend_from_reader(buf)?;
+
+    index.write(file)?;
+
+    Ok(())
+}
+
+async fn download_full_index(
+    cache: Arc<Cacher<FetchSymstoreIndexSegment>>,
+    downloader: Arc<DownloadService>,
+    source_config: Arc<HttpSourceConfig>,
+    scope: Scope,
+    last_id: u32,
+    file: &mut File,
+) -> CacheContents {
+    let mut futures = FuturesUnordered::new();
+    for i in 1..=last_id {
```

Comment:
Running ~200 downloads concurrently here results in some of them failing because of rate limiting or connection timeouts. Any suggestions for how to remedy this would be welcome.

Reply:
If you set this up as a `futures::Stream` you can specify the max concurrency to use with `buffer_unordered(N)`